### PR TITLE
feat: SEPA and BACS bank transfers through stripe

### PR DESCRIPTION
**Pull request by Sangamesh26**
## Type of Change

- [x] New feature

## Description

Add features SEPA and BACS bank transfers.

API contract for SEPA in hyperswitch :
```
   “payment_method”: “bank_transfer”
  “payment_method_type”: ”sepa”,
  "payment_method_data": {
    "bank_transfer": {
      "sepa_bank_transfer":{
        "billing_details": {
          "email": "johndoe@example.com",
          "name": "John"
        }
        "country": "FR"
    }
  }
},
```

API contract for BACS in hyperswitch :
```
  “payment_method”: “bank_transfer”
  “payment_method_type”: ”bacs”,
  "payment_method_data": {
   "bank_transfer": {
    "bacs_bank_transfer":{
        "billing_details": {
          "email": "johndoe@example.com",
          "name": "John"
        }
    }
  }
 },
```

Steps followed for SEPA and BACS :
1.  PaymentIntent call to stripe with customer_id 
2. We get the bank instructions in next action
3. The bank instruction are sent in confirm call of hyperswitch 
4. Hyperswitch receives webhook for the status

## Motivation and Context

New payment method , Bank transfer BACS and SEPA is implemented

## How did you test it?

Manual

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code

### CODE REVIEW

File: crates/api_models/src/payments.rs [hunk 1]
```diff
@@ -1040,9 +1083,16 @@ pub struct AchTransfer {
 }
 
 #[derive(Clone, Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
-pub struct ReceiverDetails {
-    pub amount_received: i64,
-    pub amount_charged: i64,
+#[serde(untagged)]
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:1086]: Is this enum used for `Serialize` or `Deserialize`?

File: crates/router/src/connector/stripe/transformers.rs [hunk 2]
```diff
@@ -280,6 +280,68 @@ pub struct StripeBankRedirectData {
...
+}
+
+#[derive(Debug, Eq, PartialEq, Serialize)]
+pub struct BacsBankTransferData {
+    #[serde(rename = "payment_method_data[type]")]
+    pub payment_method_data_type: StripePaymentMethodType,
+    #[serde(rename = "payment_method_data[billing_details][email]")]
+    pub email: Email,
+    #[serde(rename = "payment_method_data[billing_details][name]")]
+    pub name: String,
```
Comment (Narayanbhat166) [inline crates/router/src/connector/stripe/transformers.rs:296]: These fields are already there in 
https://github.com/juspay/hyperswitch/blob/a6cc8e08c560bec2e59c7c7290a6382d1611e8df/crates/router/src/connector/stripe/transformers.rs#L1723 can reuse the same

File: crates/api_models/src/payments.rs [hunk 3]
```diff
@@ -694,13 +694,14 @@ pub enum BankRedirectData {
...
-    pub billing_details: AchBillingDetails,
+pub struct AchBillingDetails {
+    pub email: Email,
 }
 
 #[derive(Debug, Clone, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
-pub struct AchBillingDetails {
+pub struct SepaAndBacsBillingDetails {
     pub email: Email,
+    pub name: String,
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:704]: Make this field as a secret

File: crates/api_models/src/payments.rs [hunk 4]
```diff
@@ -1039,10 +1050,34 @@ pub struct NextAction {
...
+    SepaBankInstructions(SepaBankTransferInstructions),
+    BacsBankInstructions(BacsBankTransferInstructions),
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct SepaBankTransferInstructions {
+    pub account_holder_name: String,
+    pub bic: String,
+    pub country: String,
+    pub iban: String,
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:1071]: These fields should be secret too.

File: crates/api_models/src/payments.rs [hunk 5]
```diff
@@ -1039,10 +1050,34 @@ pub struct NextAction {
...
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct BacsBankTransferInstructions {
+    pub account_holder_name: String,
+    pub account_number: String,
+    pub sort_code: String,
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:1078]: These should be a secret too

File: crates/router/src/connector/stripe/transformers.rs [hunk 6]
```diff
@@ -286,6 +286,64 @@ pub struct StripeBankRedirectData {
...
+    #[serde(rename = "owner[email]")]
+    pub email: Email,
+}
+
+#[derive(Debug, Eq, PartialEq, Serialize)]
+pub struct BacsBankTransferData {
+    #[serde(rename = "payment_method_data[type]")]
+    pub payment_method_data_type: StripePaymentMethodType,
+    #[serde(flatten)]
+    pub billing_address: StripeBillingAddress,
```
Comment (Narayanbhat166) [inline crates/router/src/connector/stripe/transformers.rs:300]: billing_address is already there in `PaymentIntentRequest`, you can use the same

File: crates/router/src/connector/stripe/transformers.rs [hunk 7]
```diff
@@ -1513,16 +1662,21 @@ impl<F, T>
...
-            });
+        let redirect_data = item.response.next_action.to_owned();
+        let redirection_data =
+            if let Some(redirection) = redirect_data {
+                match redirection {
+                    StripeNextActionResponse::DisplayBankTransferInstructions(_) => None,
+                    _ => Some(services::RedirectForm::from((
+                        redirection.get_url().ok_or(
+                            errors::ConnectorError::MissingRequiredField { field_name: "url" },
+                        )?,
```
Comment (Narayanbhat166) [inline crates/router/src/connector/stripe/transformers.rs:1673]: Cannot raise this error when handling the response. We should handle the redirection if it is provided by the connector in response, if not this should be sent as none

File: crates/router/src/connector/stripe/transformers.rs [hunk 8]
```diff
@@ -1541,12 +1695,37 @@ impl<F, T>
...
+                                amount_remaining: response.amount_remaining,
+                            },
+                        })
+                    }
+                    _ => None,
+                }).map(|response| {
+                     common_utils::ext_traits::Encode::<SepaAndBacsBankTransferInstructions>::encode_to_value(
+                &response,
+            )
+            .change_context(errors::ConnectorError::RequestEncodingFailed)
```
Comment (Narayanbhat166) [inline crates/router/src/connector/stripe/transformers.rs:1720]: This should be response handling failed

File: crates/api_models/src/payments.rs [hunk 9]
```diff
@@ -1039,10 +1050,34 @@ pub struct NextAction {
...
-    pub ach_credit_transfer: AchTransfer,
+    #[serde(flatten)]
+    pub bank_transfer_instructions: BankTransferInstructions,
     pub receiver: ReceiverDetails,
 }
 
+#[derive(Clone, Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum BankTransferInstructions {
+    AchCreditTransfer(AchTransfer),
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:1061]: Use Box for all struct variants.

File: crates/router/src/connector/stripe/transformers.rs [hunk 10]
```diff
@@ -332,24 +390,12 @@ pub struct BankTransferData {
...
 #[serde(untagged)]
-pub enum StripePaymentMethodData {
-    Card(StripeCardData),
-    PayLater(StripePayLaterData),
-    Wallet(StripeWallet),
-    BankRedirect(StripeBankRedirectData),
-    BankDebit(StripeBankDebitData),
-    AchBankTransfer(BankTransferData),
+pub enum StripeBankTransferData {
+    AchBankTransfer(AchBankTransferData),
```
Comment (jarnura) [inline crates/router/src/connector/stripe/transformers.rs:396]: Use Box for all struct variants

File: crates/router/src/connector/stripe/transformers.rs [hunk 11]
```diff
@@ -332,24 +390,12 @@ pub struct BankTransferData {
...
-pub struct StripeAchSourceRequest {
-    #[serde(rename = "type")]
-    pub transfer_type: StripePaymentMethodType,
-    #[serde(rename = "owner[email]")]
-    pub email: Email,
-    pub currency: String,
-}
-
 #[derive(Debug, Eq, PartialEq, Serialize)]
 #[serde(untagged)]
```
Comment (jarnura) [inline crates/router/src/connector/stripe/transformers.rs:394]: Add a comment which says, when de-serialise is added then untagged should be removed.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/api_models/src/payment_methods.rs [hunk 12]
```diff
@@ -205,6 +205,7 @@ pub struct ResponsePaymentMethodTypes {
 
     /// The Bank debit payment method information, if applicable for a payment method type.
     pub bank_debits: Option<BankDebitTypes>,
+    pub bank_transfers: Option<BankTransferTypes>,
```
Comment (jarnura) [inline crates/api_models/src/payment_methods.rs:208]: Add Doc comments

File: crates/api_models/src/payment_methods.rs [hunk 13]
```diff
@@ -217,6 +218,11 @@ pub struct ResponsePaymentMethodsEnabled {
     pub payment_method_types: Vec<ResponsePaymentMethodTypes>,
 }
 
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, ToSchema, PartialEq, Eq)]
+pub struct BankTransferTypes {
+    pub eligible_connectors: Vec<String>,
```
Comment (jarnura) [inline crates/api_models/src/payment_methods.rs:223]: Add Doc comments

**Review by jarnura — APPROVED** [review]

**Review by jarnura — CHANGES_REQUESTED** [review]
Add doc comments.

File: crates/router/src/connector/stripe/transformers.rs [hunk 14]
```diff
@@ -1916,14 +2125,16 @@ impl TryFrom<&types::PaymentsPreProcessingRouterData> for StripeAchSourceRequest
...
-                })?,
+            payment_method_data: AchBankTransferData {
+                email: item
+                    .request
+                    .email
+                    .clone()
+                    .get_required_value("email")
+                    .change_context(errors::ConnectorError::MissingRequiredField {
+                        field_name: "email",
+                    })?,
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe/transformers.rs:2136]:
```suggestion
                email: item.request.get_email()?,
```

File: crates/router/src/connector/stripe/transformers.rs [hunk 15]
```diff
@@ -2326,35 +2560,60 @@ impl
...
+                .response
+                .parse_struct("ChargesResponse")
+                .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
+
+            types::RouterData::try_from(types::ResponseRouterData {
+                response,
+                data: data.clone(),
+                http_code: res.status_code,
+            })
+            .change_context(errors::ConnectorError::ResponseHandlingFailed)
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe/transformers.rs:2601]:
```suggestion
```
not required

File: crates/router/src/connector/stripe/transformers.rs [hunk 16]
```diff
@@ -2326,35 +2560,60 @@ impl
...
+                .response
+                .parse_struct("PaymentIntentResponse")
+                .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
+
+            types::RouterData::try_from(types::ResponseRouterData {
+                response,
+                data: data.clone(),
+                http_code: res.status_code,
+            })
+            .change_context(errors::ConnectorError::ResponseHandlingFailed)
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe/transformers.rs:2614]:
```suggestion
```
not required

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

**Comment by jagan-jaya** [comment]
@Sangamesh26 Please run redirection tests once you are done with the changes

**Review by Narayanbhat166 — APPROVED** [review]

**Review by HyperSwitchers — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Pull request merged**